### ISSUE TITLE
Metal greedy GPU argmax fold ignores the repetition penalties, which are on by default

### ISSUE THREAD

**Issue by antonellof** [antonellof/ferrox#170]
## What happens

On Metal with `--ngl 99` at `--temp 0`, the greedy GPU argmax fold
computes the argmax of the RAW logits on the device. The host sampler
it replaces applies the repetition, frequency and presence penalties
first. `--repeat-penalty` defaults to **1.1** in this project
(llama.cpp's default is 1.0), so on a plain `--temp 0` run the penalty
is live and the fold silently drops it.

Measured on an M2 Pro, `Llama-3.2-3B-Instruct-Q4_K_M --ngl 99
--temp 0 --no-cnv`, 32 tokens, same binary:

| fold | `--repeat-penalty` | completion continues |
|---|---|---|
| on  | 1.1 (default) | `It is a universal force that affects everything...` |
| off | 1.1 (default) | `It is caused by the curvature of spacetime...` |
| on  | 1.0 | `It is a universal force that affects everything...` |
| off | 1.0 | `It is a universal force that affects everything...` |

At 1.0 the two paths agree exactly. At the default they do not. Both
are deterministic. So the fold is not equivalent to the sampler it
replaces whenever a penalty is non-neutral, which is the default.

## Why nothing catches it

`SamplingParams::greedy_equals_argmax`
(`ferrox-models/src/sampling/params.rs:181`) is the single predicate
that decides whether a device argmax may stand in for the chain:

```rust
pub fn greedy_equals_argmax(&self) -> bool {
    let runs = |step| self.sampler_order.steps().contains(&step);
    !(runs(ChainStep::Xtc) && self.xtc_can_fire())
        && !(runs(ChainStep::TypP) && self.typical_p < 1.0)
        && !(runs(ChainStep::Dry) && self.dry.is_enabled())
}
```

It tests XTC, typical-p and DRY. It does not test the penalties, which
move logits exactly the way DRY does. Two structures that must agree
about when an argmax is the answer, with nothing enforcing it: the
repo's dominant defect shape.

Both gates that consume it are affected:
`ferrox_cli::run`'s `needs_vocab_logits` and
`ferrox_server::generate::greedy_gpu_fold_allowed`.

## Why it matters

This is a wrong answer on the MAIN thread, in the default
configuration, on the backend most users of this project run. It is
not the thread-affinity issue (#166); #166 only made it visible,
because moving the step to a rayon worker turned the fold off and the
two answers stopped agreeing.

## What would settle it

Add the penalties to `greedy_equals_argmax`, and a test that walks the
chain steps and asserts every step that can move or remove a logit is
represented in the predicate, so the next sampler added cannot be
forgotten there. The predicate is the right place: it is already
shared by the CLI gate, the server gate and the sampler's own greedy
fast path.

## Reachable from untrusted input

No. Correctness only.
